classes/setup_item_processor.php

```diff
@@ -81,78 +85,192 @@ public function process(setup_item $item) {
+                // Check if pagetypepattern contains a valid module.
+                if (array_key_exists($currentmoduletype, $moduletypes)) {
 
-                    if (!key_exists($item->get_blockname(), $page->blocks->get_addable_blocks())) {
-                        $this->logger->log_message('Skipped adding new instance of ' . $item->get_blockname()
-                            . ' The block is not addable for the course page of course ' . $course->id);
+                    $moduleinstances = get_coursemodules_in_course($currentmoduletype, $course->id);
+
+                    if (empty($moduleinstances)) {
+                        $this->logger->log_message('Skipped adding new instance of ' . $item->get_blockname() . '. '
+                          . ' The course with id ' . $course->id . ' does not contain an instance of the '
+                          . $currentmoduletype . ' module.');
                         continue;
                     }
 
-                    $this->add_block($page, $item);
-                    $this->logger->log_message('Added a new instance of ' . $item->get_blockname() . ' to course ' . $course->id);
-                } else {
+                    foreach ($moduleinstances as $moduleid => $module) {
+
+                        $context = \context_module::instance($moduleid);
+
+                        // Create page.
+                        $page = new \moodle_page();
+                        $page->set_context($context);
+                        $page->set_cm($module, $course);
+                        $page->set_pagetype($item->get_pagetypepattern());
+
+                        if (!$page->blocks instanceof block_manager) {
+                            // TODO: disable whole plugin if block manager is not overridden.
+                            throw new \coding_exception(
+                              'Terminate processing. Block manager class is not configured in config.php');
+                        }
+
+                        try {
+                            // Create block instances.
+                            $page->blocks->add_region($item->get_region(), false);
+                            $page->blocks->load_blocks(true);
+                            $page->blocks->create_all_block_instances();
+
+                            // Check if block exists.
+                            $blockexist = $page->blocks->is_block_present($item->get_blockname());
+
+                            if (!$blockexist) {
+                                $blockaddable = key_exists($item->get_blockname(), $page->blocks->get_addable_blocks());
+                                if (!$blockaddable) {
+                                    $this->logger->log_message('Skipped adding new instance of ' . $item->get_blockname()
+                                        . ' The block is not addable for module' . $moduleid . 'in course ' . $course->id);
```

Comment:
This generates "Skipped adding new instance of online_users The block is not addable for module19in course 4". Needs spaces between  _module19in_

Reply:
This has been resolved